torch/csrc/jit/codegen/cuda/runtime/tensor_core.cu

```diff
@@ -0,0 +1,410 @@
+struct MatrixCoordinate {
+  nvfuser_index_t x;
+  nvfuser_index_t y;
+
+  __device__ MatrixCoordinate(nvfuser_index_t x_, nvfuser_index_t y_)
+      : x(x_), y(y_) {}
+  __device__ MatrixCoordinate(const MatrixCoordinate& other)
+      : x(other.x), y(other.y) {}
+
+  // Helper function to index shared mem,
+  //   if we decided to wrap smem pointers in a class should probably move this
+  //   there.
+  __device__ inline nvfuser_index_t linearize(const MatrixCoordinate& stride) {
+    return x * stride.x + y * stride.y;
+  }
+
+  __device__ inline nvfuser_index_t linearize(
+      nvfuser_index_t stride_x,
+      nvfuser_index_t stride_y) {
+    return x * stride_x + y * stride_y;
+  }
+};
+
+// Functional helper for composite swizzle pattern
+// This utility is used in composite swizzles, which could be
+//   useful in building configurable gemms.
+// Example:
+//   if a gemm uses mma32x32x8 and needs to vectorize read
+//  from a 128x32 shared mem block, the swizzling can be
+//  composed as:
+//    swizzle_sequence<
+//          Smem128x32V8Swizzle, MMA32X32X8Read
+//     > ({x,y}, lane_id)
+// This pattern should be enough for allowing a range of
+//  tiling options to be configured automatically.
+template <typename Swizzle = void, typename... Swizzles>
+__device__ inline MatrixCoordinate swizzle_sequence(
+    MatrixCoordinate coord,
+    int lane_id) {
+  return Swizzle::run(swizzle_sequence<Swizzles...>(coord, lane_id), lane_id);
+}
+
+// Base definition for the static recursion.
+template <>
+__device__ inline MatrixCoordinate swizzle_sequence(
+    MatrixCoordinate coord,
+    int lane_id) {
+  return coord;
+}
+
+} // namespace mem_swizzle
+
+namespace warp_mma {
+// Macro instructions built with primitives from `mma_instruction`. Each
+//  one operates on warp level.
+
+using mem_swizzle::MatrixCoordinate;
+
+// Interleaved 32x32x8 mma tile:
+//  per thread register:
+//   A[16] x B[16] -> C[32]
+// TODO (when build out):
+//   may need to add additional specifiers for operand layouts
+class mmaM32N32K8 {
+ public:
+  // options using fp16 mma accumulate
+  __device__ static inline void run(__half* C, __half* A, __half* B) {
+    run_helper(C, A, B);
+  }
+  // options using fp32 mma accumulate
+  __device__ static inline void run(float* C, __half* A, __half* B) {
+    run_helper(C, A, B);
+  }
+
+  // Swizzling definitions:
+
+  class SmemWriteASwizzle {
+   public:
+    __device__ static inline MatrixCoordinate run(
+        const MatrixCoordinate& coord,
+        int lane_id) {
+      const nvfuser_index_t x = coord.x;
+
+      // mma row major tends to group 4 rows together
+      const nvfuser_index_t row_group_idx = x / 4;
+
+      // This implements mma884 specific operand A layout pattern
+      //   (cf. ISA 9.7.13.4)
+      //  swizzling at write store so load can be less integer heavy.
+      // switch bit 0 and bit 1 of row_group_idx
+      const nvfuser_index_t row_group_swizzled = (row_group_idx & ~0b11) |
```

Comment:
Is `nvfuser_index_t` assumed to be `int`?

Reply:
This just looks like a magic to me. Is this technique also used in CUTLASS?

Reply:
This is a macro that could be either int64_t or int.

Reply:
Yes I think similar computations are around in other libs:
[https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/layout/tensor_op_multiplicand_sm70.h#L811-L813](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/layout/tensor_op_multiplicand_sm70.h#L811-L813
)
We are not exactly the same though. I just chose ways that are more conveniently mapped to fuser pipeline.